src/dlr_relayvm.cc

```diff
@@ -0,0 +1,284 @@
+  return dtype;
+}
+
+void RelayVMModel::SetInput(const char* name, const int64_t* shape, void* input, int dim) {
+  int index = GetInputIndex(name);
+  DLDataType dtype = GetInputDLDataType(index);
+  DLTensor input_tensor;
+  input_tensor.data = input;
+  input_tensor.ctx = ctx_;
+  input_tensor.ndim = dim;
+  input_tensor.shape = const_cast<int64_t*>(shape);
+  input_tensor.strides = nullptr;
+  input_tensor.byte_offset = 0;
+  input_tensor.dtype = dtype;
+  std::vector<int64_t> arr_shape(shape, shape + dim);
+
+  tvm::runtime::NDArray input_arr = tvm::runtime::NDArray::Empty(arr_shape, dtype, ctx_);
+  input_arr.CopyFrom(&input_tensor);
+  inputs_[index] = input_arr;
+}
+
+void RelayVMModel::UpdateInputs() {
+  const int kNumArgs = GetNumInputs() + 1;
+  TVMValue values[kNumArgs];
+  int type_codes[kNumArgs];
+  auto arg_setter = tvm::runtime::TVMArgsSetter(values, type_codes);
+  arg_setter(0, "main");
+  for (int i = 0; i < inputs_.size(); i++) {
+    arg_setter(i + 1, inputs_[i]);
+  }
+
+  tvm::runtime::PackedFunc set_input = vm_module_->GetFunction("set_input");
+  tvm::runtime::TVMRetValue rv;
+  set_input.CallPacked(tvm::runtime::TVMArgs(values, type_codes, kNumArgs), &rv);
+}
+
+void RelayVMModel::Run() {
+  // Invoke inference
+  UpdateInputs();
+  tvm::runtime::PackedFunc invoke = vm_module_->GetFunction("invoke");
+  output_ref_ = invoke("main");
+  UpdateOutputs();
+}
+
+void RelayVMModel::UpdateOutputs() {
+  outputs_.resize(num_outputs_);
+  if (output_ref_->IsInstance<tvm::runtime::ADTObj>() && num_outputs_ > 1) {
+    auto adt = tvm::runtime::Downcast<tvm::runtime::ADT>(output_ref_);
+    for (int i = 0; i < adt.size(); i++) {
+      outputs_[i] = tvm::runtime::Downcast<tvm::runtime::NDArray>(adt[i]);
+    }
+  } else if (output_ref_->IsInstance<tvm::runtime::NDArray::ContainerType>() && num_outputs_ == 1) {
+    outputs_[0] = tvm::runtime::Downcast<tvm::runtime::NDArray>(output_ref_);
+  } else {
+    throw dmlc::Error("Invalid output_ref format!");
+  }
+}
+
+void RelayVMModel::GetOutput(int index, void* output) {
+  CHECK_LT(index, num_outputs_) << "Output index is out of range.";
+  auto out_array = outputs_[index];
+  DLTensor output_tensor;
+  output_tensor.data = output;
+  output_tensor.ctx = ctx_;
+  output_tensor.ndim = out_array->ndim;
+  output_tensor.dtype = out_array->dtype;
+  output_tensor.shape = out_array->shape;
+  output_tensor.strides = nullptr;
+  output_tensor.byte_offset = 0;
+  out_array.CopyTo(&output_tensor);
+}
+
+void RelayVMModel::GetOutputShape(int index, int64_t* shape) const {
+  CHECK_LT(index, num_outputs_) << "Output index is out of range.";
+  std::memcpy(shape, outputs_[index]->shape, sizeof(int64_t) * outputs_[index]->ndim);
+}
+
+void RelayVMModel::GetOutputSizeDim(int index, int64_t* size, int* dim) {
+  CHECK_LT(index, num_outputs_) << "Output index is out of range.";
+  *size = 1;
+  auto arr = outputs_[index];
+  for (int i = 0; i < arr->ndim; ++i) {
+    *size *= arr->shape[i];
```

Comment:
Can this be implemented using std:accumulate?

Reply:
👍